### PR TITLE
[Model Averaging] Post-localSGD optimizer

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62132
* __->__ #62131

Wrap `PeriodicModelAverager` as an optimizer.

Currently both the optimizer and averager require an input `params` arg, where the latter actually can read params from the optimizer wrapper. Will update averager class API in a follow-up PR.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29881465](https://our.internmc.facebook.com/intern/diff/D29881465/)

### CODE REVIEW

File: torch/distributed/optim/post_localSGD_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    def zero_grad(self):
+        self.local_optimizer.zero_grad()
+
+    # TODO
+    def state_dict(self):
+        pass
+
+    # TODO
+    def load_state_dict(self, state_dict):
```
Comment (rohan-varma) [inline torch/distributed/optim/post_localSGD_optimizer.py:88]: Shall we raise `NotImplementedError` for these for now?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -4349,6 +4350,57 @@ def _test_DistributedDataParallel_SyncBatchNorm(
...
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "Only NCCL and GLOO backend support DistributedDataParallel",
+        )
+        def test_post_localSGD_optimizer_parity(self, grad_is_view=False):
+            learning_rate = 0.03
+            period = 4
+            warmup_steps = 10
+            torch.cuda.set_device(self.rank)
+            net = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(DDP_NET).cuda(),
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4364]: Shall we test parity for more complicated model such as torchvision.models.resnet50?

Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:4364]: I am afraid this will only work if torchvision is installed.

Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:4491]: yep, if you want, you can gate it with `HAS_TORCHVISION` flag.

**Review by rohan-varma — APPROVED** [review]
lgtm

**Pull request closed without merging**